### ISSUE TITLE
DMD location inside FullDMD

### ISSUE THREAD

**Issue by francisdb** [vpinball/b2s-backglass#70]
**Feature request**

Let the b2s author indicate the location of the hole reserved for the DMD.

This would have to be a relative position so coordinates within the FullDMD window can be calculated from the actual window size.

eg: 
```
x1, y1, x2, y2
0.2, 0.2, 0.8, 0.6
FullDMDRes = 1024*768
DMDLoc = 0.2*1024, 0.2*768, 0.8*1024, 0.6*768
```
On standalone we could then automatically place the DMD window relative to the FullDMD window.

**Comment by francisdb** [comment]
In case of 2 screens the location might be in the backglass instead of the FullDMD

eg: Checkpoint (Data East 1991)

**Comment by JockeJarre** [comment]
Sounds like a cool feature... but for the [B2S Designer](https://github.com/vpinball/b2s-designer) (as a first checkpoint) ?

This way the directb2s files would get this information into it's xml structure.

**Comment by francisdb** [comment]
Feel free to move this to the appropriate repo.

**Comment by francisdb** [comment]
Regarding looking into this I guess that's not doable on Linux/Mac?

**Comment by JockeJarre** [comment]
> Feel free to move this to the appropriate repo.

We can leave it here, cause we will need a change both here and in the Designer. So another issue on the Designer to make it possible to mark the DMD position.
By the way, I checked your Table inspection tool as inspiration for my VPinballX.starter.

> Regarding looking into this I guess that's not doable on Linux/Mac?

You are right, it doesn't make it easier. I guess .NET isn't a blocker anymore, but these tools use Windows Forms for the GUI, so that isn't easy. I know there are users running the B2S Server on linux through Wine.
But it is of course possible to use a VM with VS already installed. M$ even has the VM files available for download with a license active for a couple of months.

**Comment by francisdb** [comment]
@JockeJarre A `Developer` section in the readme that explains how to build the project would be great. Same for the designer.

Not saying that I will find the time to look into this as I have other higher priority standalone issues I would like to see fixed first.

**Comment by JockeJarre** [comment]
Good point! Done!

**Comment by francisdb** [comment]
Personally I am mostly interested in the `directb2s` file and the designer as vpinball standalone ported the server code to C++ and standalone has a specific way of showing dmd's.

https://github.com/vpinball/vpinball/tree/standalone/standalone/inc/b2s

**Comment by francisdb** [comment]
ping @jsm174

**Comment by jsm174** [comment]
Sorry, I'm not fully following. So we know the position of the B2S DMD window, but not the position of the VPinMAME or FlexDMD window inside the B2S DMD window. Or am I wrong on this?

**Comment by francisdb** [comment]
We want to know from the .directb2s file what the location of this dmd hole is so we can put the dmd automatically at the correct location.

Example below

![image](https://github.com/vpinball/b2s-backglass/assets/161305/68a38873-3dd9-4263-b202-b81e9c087dbd)

**Comment by JockeJarre** [comment]
So that future version of the B2S.Server could potentially know where the DMD should be placed. Hopefully make a more seamless experience...

**Comment by toccata10** [comment]
Hi,
I'm developing the configgen on batocera for vpinball. The idea is to get everything as easy as possible for the user.
So, full DMD currently requires per table manual adjustment of the position and size. It would be great if this thing could be implemented.
I just wanted to let you know that this enhancement would be super useful for us to have automatic positioning.

**Comment by JockeJarre** [comment]
If you want to make it as easy as possible, I would inspect the images for the full DMD and grill and then find the largest rectangle within the image, to automatically define the position. Se the image examples above...
Or maybe to make it more bullet proof, have the full DMD and grill images defined as transparent where the DMD is to be positioned.

**Comment by francisdb** [comment]
The simplest way is adding 2 relative coordinates to the xml. Other heuristics will be imprecise and slow down b2s startup.

**Comment by toccata10** [comment]
I completely agree that it's best to add coordinates in the xml. I would say: x,y,width and height.

**Comment by francisdb** [comment]
For example here finding an empty square or transparent pixels won't help

![image](https://github.com/vpinball/b2s-backglass/assets/161305/8c4c79f5-bba1-4a02-a603-b6872144979c)

**Comment by JockeJarre** [comment]
yes, sounds good! The B2S.Designer needs the possibility to mark the DMD position and save it in the directb2s file.
Here is the repository: [b2s-designer](https://github.com/vpinball/b2s-designer)

**Comment by Ltek** [comment]
this is 100% needed and was quite surprised it wasnt added long ago. basic text in a location of the B2S that other apps can read... even if it need to be manually input (eg, not using the designer's GUI to find the coordinates) would be better than nothing.

**Comment by JockeJarre** [comment]
It is so good we are open source, just hack away!

**Comment by stevejones72** [comment]
Must admit this whole resizing all my DMD's has gotten a 'stale'.....

**Comment by ScreaminDemon** [comment]
For 2 screen I would make an alternate proposal. 

The objective is to remove a few complex process from BG designer workload and at the same time standardize the output. That means that only by exception does the BG designer need to do any work related to the BG Grill - semi automated. 

In the B2S designer app, BG designer only has to pay attention to the BG image. No need to set Grill/DMD dimensions or include it in the BG image. As part of the BG design process the designer only has to "call out/declare" the required Grill and B2S will insert it automatically when B2S server runs that BG. B2S server should include default Grill image of each table brand, including a few generic Grills e.g. unbranded round speaker, unbranded rectangular speaker etc. All default Grills are designed  to be same dimensions and all DMD in same position.

**Comment by francisdb** [comment]
Looks like something similar has been implemented by vpin studio.

https://github.com/syd711/vpin-studio/releases/tag/3.11.0

![image](https://github.com/user-attachments/assets/9dbbfc02-8899-4856-947d-6940fad6ace1)

**Comment by francisdb** [comment]
@Le-Syl21 wrote a tool to automate this
He is using [Open Computer Vision](https://opencv.org/get-started/) to find the hole
https://github.com/Le-Syl21/dmd-config

**Comment by superhac** [comment]
> [@Le-Syl21](https://github.com/Le-Syl21) wrote a tool to automate this He is using [Open Computer Vision](https://opencv.org/get-started/) to find the hole https://github.com/Le-Syl21/dmd-config

I can confirm I did over a dozen tables now and its been spot on.  I couldn't do it better by hand on standalone. Nice work and thanks for the great tool!

**Comment by francisdb** [comment]
To be complete, vpin studio also has this and uses a similar computer vision library

https://github.com/syd711/vpin-studio/blob/main/vpin-studio-server/src/main/java/de/mephisto/vpin/server/dmd/DMDPositionDetector.java
